doc: Added Miscellaneous section for Docker users

ShahanaFarooqui opened the pull request:
Added CLN image execution and testing hints for Docker users.

Changelog-None.

ShahanaFarooqui commented:
@chrisguida Testing if I can add you as reviewer after the comment :).

ShahanaFarooqui commented:
> @chrisguida Testing if I can add you as reviewer after the comment :).

Nope, still cannot :(

chrisguida commented:
Hey Shahana, running this on my laptop:

```
$ docker run -it --rm --platform=linux/amd64 -d --network=host -v '/usr/.lightning:/root/.lightning' -v '/usr/.bitcoin:/usr/.bitcoin' -e LIGHTNINGD_DATA=/root/.lightning -e LIGHTNINGD_NETWORK=regtest -p 9735:9735/tcp -p 9835:9835/tcp elementsproject/lightningd:latest --bitcoin-rpcconnect=0.0.0.0 --bitcoin-rpcuser=chris --bitcoin-rpcpassword=password --network=regtest --database-upgrade=true
WARNING: Published ports are discarded when using host network mode
c5b3c5accef7b3fe8ea4b53fd6ae20058180f71b1dd51b06580657c4ca7f3e52
```

I just get a hash. Nothing in `docker ps` and no error messages. What am I doing wrong? Is there a log generated anywhere?

chrisguida reviewed (changes requested):
Hey Shahana, thanks for adding this!

I think a better command might be:

`docker run -it --rm --network=host -v $HOME/.lightning/regtest:/root/.lightning -v $HOME/.bitcoin:/root/.bitcoin elementsproject/lightningd:latest --network=regtest --database-upgrade=true`

We should use the user's home dir as the bitcoin dir, because that's most likely where they're actually running bitcoind from. That way we can use the cookie instead, which allows us to leave out the rpcuser and rpcpassword stuff, which requires the user to update their bitcoind config.

The ports you included don't do anything, as lightningd listens on port 19846 by default, not 9735 or 9835, though we could perhaps add a note or another example for running on mainnet (maybe not recommendable). Also, as the warning notes, with `--network=host` turned on, the ports are ignored anyway.

The environment variables `-e LIGHTNINGD_DATA=/root/.lightning -e LIGHTNINGD_NETWORK=regtest` also don't do anything, since this is already the default datadir and `--network=regtest` is passed in later.

We should remove the `--platform` argument, as this will not work if a user runs it from something other than x86, such as ARM.

We should remove the `-d` flag, as this causes the output to disappear. I actually was unable to figure out where the output was going, and unable to debug this command at all, until I removed the `-d`. This will help users debug their setup, if they still have problems like custom stuff in their home directory, which could cause lightningd to crash, as it did for me with my `proxy` option.

We should probably try to take the `modded` flag out of the release version and make this image just use the official release binary. That way we can also remove the `--database-upgrade=true` argument.

We probably also want to add a note about how to access `lightning-cli` once the image is up.

For me, I did `docker exec -it <container-name> bash`, then `lightning-cli --regtest getinfo`.

Let me know if you need anything from me to help get this done! :)

chrisguida commented:
On second thought, the user's home directory is probably not the best place for the lightning dir, as this causes the files to become owned by root. Maybe putting it in `/usr/`is better then. Or maybe something like `/var/lib/docker`.

ShahanaFarooqui commented:
> We should use the user's home dir as the bitcoin dir, because that's most likely where they're actually running bitcoind from. That way we can use the cookie instead, which allows us to leave out the rpcuser and rpcpassword stuff, which requires the user to update their bitcoind config.
> 

Thanks for the suggestion. Fixed. 

> The ports you included don't do anything, as lightningd listens on port 19846 by default, not 9735 or 9835, though we could perhaps add a note or another example for running on mainnet (maybe not recommendable). Also, as the warning notes, with `--network=host` turned on, the ports are ignored anyway.
> 

Yup, somehow mixed up multiple different commands. Fixed.

> The environment variables `-e LIGHTNINGD_DATA=/root/.lightning -e LIGHTNINGD_NETWORK=regtest` also don't do anything, since this is already the default datadir and `--network=regtest` is passed in later.
> 

I was using them for my post-start script and forgot to remove them. Fixed.

> We should remove the `--platform` argument, as this will not work if a user runs it from something other than x86, such as ARM.
> 

We can run different platform images with the flag as long as QEMU setup is functioning correctly on your system. Rather than removing the flag completely, I added an information box explaining it.

> We should remove the `-d` flag, as this causes the output to disappear. I actually was unable to figure out where the output was going, and unable to debug this command at all, until I removed the `-d`. This will help users debug their setup, if they still have problems like custom stuff in their home directory, which could cause lightningd to crash, as it did for me with my `proxy` option.
> 

Thanks, Fixed. Accidentally left a lot of my preferred flags :).

> We should probably try to take the `modded` flag out of the release version and make this image just use the official release binary. That way we can also remove the `--database-upgrade=true` argument.
> 

Removed the `database-upgrade` config option.

> We probably also want to add a note about how to access `lightning-cli` once the image is up.
> 

Good idea! Added the section as per your suggestion.

chrisguida commented:
@ShahanaFarooqui great update, thanks!

I still think it doesn't make sense to mount the bitcoin datadir from the user's `/root` directory; bitcoind does not install there by default so users who are not very familiar with docker's mounting syntax will probably be confused as to why things aren't working. But at least the container output is visible now, so they should be able to at least debug why :)

Maybe we could add a line saying "change the left half of the `-v` parameters to where bitcoind's and lightningd's data directories are actually located on your system"?
